src/blockchain/script_sync.rs

```
@@ -314,6 +314,26 @@ impl<'a, D: BatchDatabase> State<'a, D> {
         let finished_txs = make_txs_consistent(&self.finished_txs);
         let observed_txids: HashSet<Txid> = finished_txs.iter().map(|tx| tx.txid).collect();
         let txids_to_delete = existing_txids.difference(&observed_txids);
+
+        // Ensure `last_active_index` does not decrement database's current state.
+        let index_updates = self
+            .last_active_index
+            .iter()
+            .filter_map(|(keychain, index)| {
+                let db_index = self
+                    .db
+                    .get_last_index(*keychain)
+                    .unwrap_or(None)
```

I think here ideally we should propagate the error instead of "downgrading" it to a `None` (otherwise again we may overwrite the last derivation index accidentally).

On the other end, I'm not sure what would be the best way to do it: the only thing I could think of would be to replace this `filter_map` with just a map, and return a `Result` in the closure so that you can collect a `Result<Vec<_>>` and propagate the error with `?`. Then you'll have to filter out the `None`s before iterating the list.

---

Is this satisfactory?

https://github.com/bitcoindevkit/bdk/blob/6f6d923abcd7b5382db3f45ab97b5eb8640f5f0c/src/blockchain/script_sync.rs#L318-L341

---

Continued here: https://github.com/bitcoindevkit/bdk/pull/653#discussion_r916074702